numpy/linalg/umath_linalg.cpp

```diff
@@ -4748,10 +4741,30 @@ PyMODINIT_FUNC PyInit__umath_linalg(void)
     PyDict_SetItemString(d, "_ilp64", Py_False);
 #endif
 
-#if Py_GIL_DISABLED
-    // signal this module supports running with the GIL disabled
-    PyUnstable_Module_SetGIL(m, Py_MOD_GIL_NOT_USED);
+    return 0;
+}
+
+static struct PyModuleDef_Slot _umath_linalg_slots[] = {
+    {Py_mod_exec, (void*)_umath_linalg_exec},
+#if PY_VERSION_HEX >= 0x030c00f0  // Python 3.12+
+    {Py_mod_multiple_interpreters, Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED},
+#endif
+#if PY_VERSION_HEX >= 0x030d00f0  // Python 3.13+
+    // signal that this module supports running without an active GIL
+    {Py_mod_gil, Py_MOD_GIL_NOT_USED},
 #endif
+    {0, NULL},
+};
+
+static struct PyModuleDef moduledef = {
+    PyModuleDef_HEAD_INIT,  /* m_base */
+    "_umath_linalg",        /* m_name */
+    NULL,                   /* m_doc */
+    0,                      /* m_size */
```

Comment:
For now, in a future PR this can be changed.
```suggestion
    -1,                      /* m_size */
```

Reply:
See also my note in https://github.com/numpy/numpy/pull/29025#discussion_r2101682505. Briefly, I was mistaken in my understanding of the documentation, `-1` is only valid for single-phase initialisation, and is forbidden for multi-phase.

The safest thing reinitialisation-wise would be to use the [multiple interpreter opt-out](https://docs.python.org/3/howto/isolating-extensions.html#opt-out-limiting-to-one-module-object-per-process) suggested by CPython. This would enforce that each extension module is only imported & initialised once per process.

However, this causes ``test_reloading.py::test_full_reimport()`` to fail. Is that an acceptable trade-off? Multi-phase init (and at some point, isolated modules) opens the door to proper support for reloading/reimporting `numpy`.

cc @encukou re the best way to do a gradual migration to multi-phase


In the case of `umath_linalg`, we might be fine globals-wise, but `multiarray` will definitely need a solution here.

Reply:
I've [just updated](https://github.com/python/cpython/pull/134141) the HOWTO, see [here](https://cpython-previews--134141.org.readthedocs.build/en/134141/howto/isolating-extensions.html#opt-out-limiting-to-one-module-object-per-process) before it gets to main docs.

Basically, to allow re-import, clear the "loaded" flag in `PyModuleDef`'s `m_clear`. Note that this *will* do an actual reload of the module (sans C globals), that is, *exec* will run. (With single-phase, the module dict contents are stashed away and reused. You could emulate *that*, but )

The PyMutex stuff in the updated HOWTO is 3.13+ non-limited API, but isn't necessary in the medium term (on CPython you're protected by the GIL; even free-threading builds use an import lock which isn't likely to go away). For the long-term (or alternate CAPI implementations) you could add no-op shims.

The `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED` flag will only allow loading in the main interpreter; that's redundant if you're limited to one module per process.

Reply:
> However, this causes test_reloading.py::test_full_reimport() to fail. Is that an acceptable trade-off?

I am not opposed to breaking the test/changing the behavior.  The ask would be to summarize how things change for:
* Potential `subinterpreter` users (I _think_ that just doesn't work at all anymore, but not sure).
* Users for `importlib.reload()`, currently they get a warning that this is a bad idea.  A full-blown error may not be so bad, though.

Reply:
> Potential `subinterpreter` users (I _think_ that just doesn't work at all anymore, but not sure).

As far as I can tell, importing `numpy` simply fails in a subinterpreter, and continues to do so with the explicit opt-out.

```pycon
$ uvx --with numpy --python 3.13 python
Python 3.13.3+ (main, May 20 2025, 08:56:23) [GCC 13.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import _interpreters
>>> # first import in an interpreter with a 'legacy' shared GIL
>>> c = _interpreters.new_config('legacy')  
>>> id1 = _interpreters.create(c, reqrefs=True)
>>> _interpreters.incref(id1)  # bookkeeping
>>> _interpreters.exec(id1, 'import sys', restrict=True)  # stdlib works
>>> _interpreters.exec(id1, 'import numpy', restrict=True)
namespace(
  type=namespace(
    __name__='RuntimeError',
    __qualname__='RuntimeError',
    __module__='builtins'
  ), 
  msg='CPU dispatcher tracer already initlized',
  formatted='RuntimeError: CPU dispatcher tracer already initlized',
  errdisplay='Traceback (most recent call last):\n  File "<string>", line 1, in <module>\n'
             '[...SNIP...]\n'
             '    from numpy._core._multiarray_umath import (\n'
             '        add_docstring,  _get_implementing_args, _ArrayFunctionDispatcher)\n'
             'RuntimeError: CPU dispatcher tracer already initlized'
)
>>> # now import in an isolated interpreter with its own GIL
>>> id2 = _interpreters.create(reqrefs=True)
>>> _interpreters.exec(id2, 'import sys', restrict=True)
>>> print(_interpreters.exec(id2, 'import numpy', restrict=True))
double free or corruption (out)
$  # the REPL has crashed!
```

> Users for `importlib.reload()`, currently they get a warning that this is a bad idea. A full-blown error may not be so bad, though.

Reloading currently works with a UserWarning, unchanged with the explict opt-out.

-----

The observable change I've found is clearing `sys.modules` and re-importing. Previously, it emitted `UserWarning: The NumPy module was reloaded (imported a second time). This can in some cases result in small but subtle issues and is discouraged.`, but would now raise `ImportError: cannot load module more than once per process`.

Reply:
@seberg fab! There's already a test for `importlib.reload()` in `test_numpy_reloading()`, so I've changed this test to assert that an `ImportError` is raised. Should I remove it entirely instead?

Reply:
I don't care, but it may be good to add a release note fragment about this somewhere.

Reply:
Done, added a brief note about the changes.

Reply:
For completeness, could you please check that [`%autoreload`](https://ipython.org/ipython-doc/3/config/extensions/autoreload.html#module-IPython.extensions.autoreload) in IPython/Jupyter isn't affected?

Reply:
Awesome, thanks for verifying. And yes, the implementation is shared.